### PR TITLE
[Feature] loss function

### PR DESCRIPTION
**Pull request by menjarleev**
…FB15k using different losses. modify forward_test logic. need to merge with edge weight

*Issue #, if available:*
1. no comments, need to fix tmr
2. no abstract class for MXNet
3. need to merge with edge weight implementation
*Description of changes:*
1. isolate loss to a separate class
2. add hinge, logistic loss support
3. add pair-wise loss support
4. update forward_test logic

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Related issue: #159

### CODE REVIEW

File: python/dglke/models/base_loss.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        if self.pairwise and self.loss_genre == 'Ranking':
+            raise ValueError('Ranking loss cannot be applied to pairwise loss function')
+
+    def get_pos_loss(self, pos_score):
+        """ Predict loss for positive labels
+
+        Parameters
+        ----------
+        pos_score : tensor
+                    Score calculated from positive triples
```
Comment (classicsong) [inline python/dglke/models/base_loss.py:83]: The indent should be 4 spaces comparing to the last line.

File: python/dglke/models/base_loss.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+        Returns
+        -------
+        tensor
+                positive loss calculated with specific loss criterion
```
Comment (classicsong) [inline python/dglke/models/base_loss.py:88]: indent

File: python/dglke/models/base_loss.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        """
+        pass
+
+    def get_neg_loss(self, neg_score):
+        """ Predict loss for negative triples
+
+        Parameters
+        ----------
+        neg_score: tensor
+                   Score calculated from positive triples
```
Comment (classicsong) [inline python/dglke/models/base_loss.py:98]: indent

File: python/dglke/models/base_loss.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+        Returns
+        -------
+        tensor
+                Negative loss calculated with specific loss criterion
```
Comment (classicsong) [inline python/dglke/models/base_loss.py:103]: indent

File: python/dglke/models/base_loss.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        L_{adv\_neg}-> adversarial weighed negative loss
+        L_{neg} : negative loss
+        f : score function
+        t_i^- : negative sample for triple i
+        T : adversarial temperature (hyper-parameter)
+
+        Parameters
+        ----------
+        pos_score : tensor
+                    Score calculated from positive triples
```
Comment (classicsong) [inline python/dglke/models/base_loss.py:134]: indent

File: python/dglke/models/base_loss.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        neg_score : tensor
+                    Score calculated from negative triples
```
Comment (classicsong) [inline python/dglke/models/base_loss.py:136]: indent

File: python/dglke/models/general_models.py [hunk 7]
```diff
@@ -438,31 +444,33 @@ def forward_test(self, pos_g, neg_g, logs, gpu_id=-1):
...
+            mask = filter_bias != -1
+
         # To compute the rank of a positive edge among all negative edges,
         # we need to know how many negative edges have higher scores than
         # the positive edge.
-        rankings = F.sum(neg_scores >= pos_scores, dim=1) + 1
-        rankings = F.asnumpy(rankings)
         for i in range(batch_size):
-            ranking = rankings[i]
+            if self.args.eval_filter:
```
Comment (classicsong) [inline python/dglke/models/general_models.py:469]: Can you profile the speed of this implementation and the previous one?

File: python/dglke/models/mxnet/loss.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        pass
+
+class LogisticLoss(BaseLogisticLoss):
+    def __init__(self, args):
+        assert False, 'LogisticLoss is not implemented'
+
+    def __call__(self, score, label):
+        pass
+
+class LogsigmoidLoss(BaseLogsigmoidLoss):
```
Comment (classicsong) [inline python/dglke/models/mxnet/loss.py:17]: We should provide logsigmoid loss for mxnet. Otherwise MXNet will not work.

File: python/dglke/models/mxnet/loss.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        """
+        pass
+
+    def _get_neg_loss(self, neg_score):
+        """ Predict loss for negative triples
+
+        Parameters
+        ----------
+        neg_score: tensor
+                   Score calculated from positive triples
```
Comment (classicsong) [inline python/dglke/models/mxnet/loss.py:58]: indent

File: python/dglke/models/mxnet/loss.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+        Returns
+        -------
+        tensor
+                Negative loss calculated with specific loss criterion
```
Comment (classicsong) [inline python/dglke/models/mxnet/loss.py:63]: indent

File: python/dglke/models/mxnet/loss.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    def get_total_loss(self, pos_score, neg_score):
+        """ Calculate total loss for a batch of positive triples and negative triples. The total loss can be
+            point-wise and pairwise. For pairwise, it is average of the relative loss from positive score to negative
+            score. For point-wise, it can be average of the positive loss and negative loss or negative loss
+            weighted by its negative score and adversarial_temperature.
+
+        Parameters
+        ----------
+        pos_score : tensor
+                    Score calculated from positive triples
```
Comment (classicsong) [inline python/dglke/models/mxnet/loss.py:76]: indent

File: python/dglke/models/mxnet/loss.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        neg_score : tensor
+                    Score calculated from negative triples
```
Comment (classicsong) [inline python/dglke/models/mxnet/loss.py:78]: indent

File: python/dglke/models/pytorch/loss.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        super(LogsigmoidLoss, self).__init__(args)
+
+    def __call__(self, score: th.Tensor, label):
+        return - logsigmoid(label * score)
+
+
+class LossGenerator(BaseLossGenerator):
+    def __init__(self, args):
+        super(LossGenerator, self).__init__(args)
+        if self.loss_genre == 'Hinge':
```
Comment (classicsong) [inline python/dglke/models/pytorch/loss.py:44]: We do not need loss_genre.

File: python/dglke/models/pytorch/loss.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+class LogsigmoidLoss(BaseLogsigmoidLoss):
+    def __init__(self, args):
+        super(LogsigmoidLoss, self).__init__(args)
+
+    def __call__(self, score: th.Tensor, label):
+        return - logsigmoid(label * score)
+
+
+class LossGenerator(BaseLossGenerator):
+    def __init__(self, args):
```
Comment (classicsong) [inline python/dglke/models/pytorch/loss.py:42]: list all the parameters and give docs for each parameters. This func can be reused in KE Python API.

File: python/dglke/models/pytorch/loss.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        return self.loss_criterion(neg_score, self.neg_label)
+
+    def get_total_loss(self, pos_score, neg_score):
+        log = {}
+        if self.pairwise:
+            pos_score = pos_score.unsqueeze(-1)
+            loss = th.mean(self.loss_criterion(pos_score - neg_score, 1))
+            log['loss'] = get_scalar(loss)
+            return loss, log
+        pos_loss = self._get_pos_loss(pos_score)
```
Comment (classicsong) [inline python/dglke/models/pytorch/loss.py:72]: Add one empty line here.

File: python/dglke/models/pytorch/loss.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        neg_loss = th.mean(neg_loss)
+        pos_loss = th.mean(pos_loss)
+        loss = (neg_loss + pos_loss) / 2
+        log['pos_loss'] = get_scalar(pos_loss)
+        log['neg_loss'] = get_scalar(neg_loss)
+        log['loss'] = get_scalar(loss)
+        return loss, log
+
+
+
```
Comment (classicsong) [inline python/dglke/models/pytorch/loss.py:90]: Too many empty lines at the end.

File: python/dglke/models/general_models.py [hunk 17]
```diff
@@ -493,13 +501,8 @@ def forward(self, pos_g, neg_g, gpu_id=-1):
         """
         pos_g.ndata['emb'] = self.entity_emb(pos_g.ndata['id'], gpu_id, True)
         pos_g.edata['emb'] = self.relation_emb(pos_g.edata['id'], gpu_id, True)
-
         self.score_func.prepare(pos_g, gpu_id, True)
-
         pos_score = self.predict_score(pos_g)
-        pos_score = logsigmoid(pos_score)
-        if self.has_edge_importance:
```
Comment (classicsong) [inline python/dglke/models/general_models.py:501]: Where do you handle edge importance?
We should support this feature.
The importance weight is applied to both pos and neg score in a pairwise manner.

**Comment by classicsong** [comment]
Give some comments.
BTW, nice work.

File: python/dglke/models/base_loss.py [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        self.neg_adversarial_sampling = args.neg_adversarial_sampling
+        if self.neg_adversarial_sampling:
+            self.adversarial_temperature = args.adversarial_temperature
+        else:
+            self.adversarial_temperature = 0
+        self.loss_genre = args.loss_genre
+        self.neg_label = args.neg_label
+        if self.pairwise == self.neg_adversarial_sampling == True:
+            raise ValueError('loss cannot be pairwise and adversarial sampled')
+        if self.pairwise and self.loss_genre == 'Ranking':
```
[the remark this answers is not in the document]
Comment (menjarleev) [reply python/dglke/models/base_loss.py:74]: BCELoss or other loss subclasses do not have attr pairwise. Thus this value error might only be placed here.

File: python/dglke/models/mxnet/loss.py [hunk 19]
```diff
@@ -1,87 +1,45 @@
...
-
-        Returns
-        -------
-        tensor
-                Negative loss calculated with specific loss criterion
-        """
+    def _get_neg_loss(self, neg_score, edge_weight):
         pass
 
-    def get_total_loss(self, pos_score, neg_score):
```
[the remark this answers is not in the document]
Comment (menjarleev) [reply python/dglke/models/mxnet/loss.py]: I've put one in BaseLossGenerator. Is it necessary to do that again in the pytorch version?

File: docs/source/train.rst [hunk 20]
```diff
@@ -121,17 +121,17 @@ The command line provides the following arguments:
...
-    The loss functions provided by DGL-KE.
+  - ``--loss_genre {Logistic, Hinge, Logsigmoid, BCE}``
+    The loss functions provided by DGL-KE. Default loss is Logistic
 
   - ``--margin`` or ``-m``
     The margin value for hinge loss.
 
   - ``--neg_label {0, -1}``
-    The label for negative sample. Normally use -1 for Logsigmoid, Logistic, etc.
+    The label for negative sample is related to loss function. -1 for Logsigmoid, Logistic, 0 for BCE.
```
[the remark this answers is not in the document]
Comment (menjarleev) [reply docs/source/train.rst:131]: yep

File: python/dglke/models/pytorch/loss.py [hunk 21]
```diff
@@ -8,69 +8,75 @@
...
         if self.pairwise:
             pos_score = pos_score.unsqueeze(-1)
-            loss = th.mean(self.loss_criterion(pos_score - neg_score, 1))
+            loss = th.mean(self.loss_criterion((pos_score - neg_score) * edge_weight, 1))
             log['loss'] = get_scalar(loss)
             return loss, log
-        pos_loss = self._get_pos_loss(pos_score)
-        neg_loss = self._get_neg_loss(neg_score)
+        pos_loss = self._get_pos_loss(pos_score, edge_weight)
+        neg_loss = self._get_neg_loss(neg_score, edge_weight)
```
[the remark this answers is not in the document]
Comment (menjarleev) [reply python/dglke/models/pytorch/loss.py:79]: ```edge_weight``` has no influence on adversarial sampling as it is computed based on '''neg_score''' not  '''neg_loss'''. Only '''neg_loss''' is weighed by '''edge_weight'''.

File: python/dglke/models/mxnet/loss.py [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+        pass
+
+class BCELoss(BaseBCELoss):
+    def __init__(self):
+        assert False, 'BCELoss is not implemented'
+
+    def __call__(self, score, label):
+        pass
+
+class LogsigmoidLoss(BaseLogsigmoidLoss):
```
Comment (classicsong) [inline python/dglke/models/mxnet/loss.py:24]: We actually can support logsigmoid right?

Comment (classicsong) [reply python/dglke/models/mxnet/loss.py:24]: Please keep it. Otherwise it is a breaking change.

**Review by classicsong — APPROVED** [review]
LGTM. Please fix the mxnet compatibility problem.

**Review by classicsong — APPROVED** [review]
LGTM
